File: services/tyre_model_service.py
import logging
from sqlalchemy import or_
from database.extensions import db
from database.unit_of_work import UnitOfWork
from policies.file_naming import uuid_filename
from werkzeug.datastructures import FileStorage
from database.models.tyre_model import TyreModel
from services.file_service import FileService, FileSaveRequest
from database.models.data_types.files import FileType, FileModel
from database.repositories.tyre_model_repository import TyreModelRepository
from domain.exceptions import DatabaseError, ModelNotFoundError, InvalidFileTypeError, InvalidFileError, FileSaveError

logger = logging.getLogger(__name__)


class TyreModelService:
    def __init__(self):
        self.repo = TyreModelRepository()
        self.file_service = FileService()

    def get_all(self, page=1, page_size=20, search="") -> (list[TyreModel], int):
        filters = None

        if search != "":
            search_term = f"%{search}%"
            filters = or_(
                TyreModel.manufacturer.ilike(search_term),
                TyreModel.model_name.ilike(search_term),
            )

        return self.repo.get_all(page=page, page_size=page_size, filters=filters)

    def get_by_id(self, id_: int) -> TyreModel:
        try:
            tyre_model = self.repo.get_by_id(id_)
        except ModelNotFoundError as e:
            logger.error(f"TyreModel with id {id_} not found: {e}")
            raise ModelNotFoundError(f"TyreModel with id {id_} not found: {e}") from e

        return tyre_model

    def create(self, dto) -> TyreModel:
        with UnitOfWork(db.session):
            try:
                tyre_model = self.repo.create(
                    manufacturer=dto.get("manufacturer", "Temp Manufacturer"),
                    model_name=dto.get("model_name", "Temp Model None"),
                    category=dto.get("category", None),
                    vehicle_type=dto.get("vehicle_type", None),
                    width_mm=dto.get("width_mm", None),
                    aspect_ratio=dto.get("aspect_ratio", None),
                    rim_diameter_inches=dto.get("rim_diameter_inches", None),
                    groove_count=dto.get("groove_count", None),
                    pattern_type=dto.get("pattern_type", None),
                    tread_pitch_length_mm=dto.get("tread_pitch_length_mm", None),
                    dataset_source=dto.get("dataset_source", None),
                    notes=dto.get("notes", None),
                )
            except DatabaseError as e:
                logger.error(f"Error creating tyre model record: {e}")
                raise DatabaseError(f"Error creating tyre model record: {e}")

        return tyre_model

    def upload_image(self, tyre_model: TyreModel, file: FileStorage) -> TyreModel:
        if not file:
            logger.error("No file provided in tyre model service")
            raise InvalidFileTypeError("No file provided")

        uuid, filename = uuid_filename(file)
        file.filename = filename

        with UnitOfWork(db.session):
            try:
                file_record = self.file_service.handle_file(
                    FileSaveRequest(
                        file=file,
                        upload_directory=f"/tyre_match/files/tyre_models/{tyre_model.id}",
                        valid_extensions=["png", "jpg", "jpeg"],
                        model=FileModel.tyre_model,
                        model_id=tyre_model.id,
                        file_type=FileType.original
                    )
                )
            except InvalidFileError as e:
                logger.error(f"Invalid file error from file service in tyre model service: {e}")
                raise FileSaveError(f"Invalid file error from file service in tyre model service: {e}") from e
            except InvalidFileTypeError as e:
                logger.error(f"Invalid file type error from file service in tyre model service: {e}")
                raise FileSaveError(f"Invalid file type error from file service in tyre model service: {e}") from e
            except (PermissionError, OSError) as e:
                logger.error(f"Permission or OS error from file service in tyre model service: {e}")
                raise FileSaveError(f"Permission or OS error from file service in tyre model service: {e}") from e
            except DatabaseError as e:
                logger.error(f"Database error from file service in tyre model service: {e}")
                raise FileSaveError(f"Database error from file service in tyre model service: {e}") from e

            tyre_model.files.append(file_record)

        return tyre_model

    def update(self, id_, dto) -> TyreModel:
        with UnitOfWork(db.session):
            try:
                tyre_model = self.repo.get_by_id(id_)
            except ModelNotFoundError as e:
                logger.error(f"TyreModel with id {id_} not found: {e}")
                raise ModelNotFoundError(f"TyreModel with id {id_} not found: {e}") from e

            try:
                allowed_fields = [
                    "manufacturer",
                    "model_name",
                    "category",
                    "vehicle_type",
                    "width_mm",
                    "aspect_ratio",
                    "rim_diameter_inches",
                    "groove_count",
                    "pattern_type",
                    "tread_pitch_length_mm",
                    "dataset_source",
                    "notes",
                ]

                update_data = {
                    k: v for k, v in dto.items()
                    if k in allowed_fields
                }

                updated_tyre_model = self.repo.update(
                    entity=tyre_model,
                    **update_data
                )
            except DatabaseError as e:
                logger.error(f"Error updating tyre model record: {e}")
                raise DatabaseError(f"Error updating tyre model record: {e}")

        return updated_tyre_model

    def delete(self, id_: int) -> bool:
        with UnitOfWork(db.session):
            try:
                res = self.repo.delete(id_)
            except ModelNotFoundError as e:
                logger.error(f"Tyre model with id {id_} not found: {e}")
                raise ModelNotFoundError(f"Tyre model with id {id_} not found")
            except DatabaseError as e:
                logger.error(f"Error deleting tyre model record: {e}")
                raise DatabaseError(f"Error deleting tyre model record: {e}")

            return res

File: services/tyre_model_service.go
package services

import (
	"fmt"
	"github.com/jinzhu/gorm"
	"github.com/labstack/gommon/random"
	"path/filepath"
	m "tyre-match-backend/db/models"
	"tyre-match-backend/db/repositories"
	"tyre-match-backend/pkg/file_storage"
)

type TyreModelDTO struct {
	Manufacturer      string
	ModelName         string
	Category          string
	VehicleType       string
	WidthMm           int
	AspectRatio       int
	RimDiameterInches int
	GrooveCount       int
	PatternType       string
	PixelsPerInch     float32
	ROITop            int
	ROILeft           int
	ROIRight          int
	ROIBottom         int
}

type TyreModelService struct {
	repo                   *repositories.TyreModelRepository
	fileRepo               *repositories.FileRepository
	fileStore              file_storage.Store
	validator              FileValidator
	imageProcessingService ImageProcessingServiceInterface
}

func NewTyreModelService(
	repo *repositories.TyreModelRepository,
	fileRepo *repositories.FileRepository,
	fileStore file_storage.Store,
	validator FileValidator,
	imageProcessingService ImageProcessingServiceInterface,
) *TyreModelService {
	return &TyreModelService{
		repo:                   repo,
		fileRepo:               fileRepo,
		fileStore:              fileStore,
		validator:              validator,
		imageProcessingService: imageProcessingService,
	}
}

func (s *TyreModelService) List(page, pageSize int, search string) ([]*m.TyreModel, int, int, int) {
	var scopes []func(db *gorm.DB) *gorm.DB

	if search != "" {
		searchTerm := fmt.Sprintf("%%%s%%", search)
		scopes = append(scopes, func(db *gorm.DB) *gorm.DB {
			return db.Where("manufacturer LIKE ? OR model_name LIKE ?", searchTerm, searchTerm)
		})
	}

	return s.repo.List(page, pageSize, scopes)
}

func (s *TyreModelService) Get(id uint) (*m.TyreModel, error) {
	tyreModel := s.repo.GetByID(id)
	if tyreModel == nil {
		return nil, NotFoundError
	}

	return tyreModel, nil
}

func (s *TyreModelService) Create(dto TyreModelDTO) (*m.TyreModel, error) {
	tyreModel := &m.TyreModel{
		Manufacturer:      dto.Manufacturer,
		ModelName:         dto.ModelName,
		WidthMm:           dto.WidthMm,
		AspectRatio:       dto.AspectRatio,
		RimDiameterInches: dto.RimDiameterInches,
		GrooveCount:       dto.GrooveCount,
		ROITop:            dto.ROITop,
		ROILeft:           dto.ROILeft,
		ROIRight:          dto.ROIRight,
		ROIBottom:         dto.ROIBottom,
		PixelsPerInch:     dto.PixelsPerInch,
		Status:            m.ProcessingStatusUploaded,
	}

	if err := s.repo.Create(tyreModel); err != nil {
		tyreModel.Status = m.ProcessingStatusFailed
		_ = s.repo.Update(tyreModel)

		return nil, err
	}

	return tyreModel, nil
}

func (s *TyreModelService) Update(id uint, dto TyreModelDTO) (*m.TyreModel, error) {
	tyreModel := s.repo.GetByID(id)
	if tyreModel == nil {
		return nil, NotFoundError
	}

	tyreModel.Manufacturer = dto.Manufacturer
	tyreModel.ModelName = dto.ModelName
	tyreModel.WidthMm = dto.WidthMm
	tyreModel.AspectRatio = dto.AspectRatio
	tyreModel.RimDiameterInches = dto.RimDiameterInches
	tyreModel.GrooveCount = dto.GrooveCount

	if err := s.repo.Update(tyreModel); err != nil {
		return nil, err
	}

	return tyreModel, nil
}

func (s *TyreModelService) Upload(id uint, file UploadedFile) (*m.TyreModel, error) {
	tyreModel := s.repo.GetByID(id)
	if tyreModel == nil {
		return nil, NotFoundError
	}
	if len(tyreModel.Images) != 0 {
		return nil, AlreadyExistsError
	}

	if err := s.validator.Validate(file); err != nil {
		return nil, InvalidUploadError
	}

	extension := filepath.Ext(file.Name)
	filename := random.String(32) + extension
	path := fmt.Sprintf("tyre-models/%v/%v", id, m.FileTypeOriginal)

	src, err := file.Open()
	if err != nil {
		return nil, err
	}
	defer src.Close()

	if err := s.fileStore.Save(path, src, filename); err != nil {
		return nil, FileStoreError
	}

	record := &m.File{
		Model:    m.FileModelTyreModel,
		ModelId:  tyreModel.ID,
		Name:     filename,
		Location: path,
		FileType: m.FileTypeOriginal,
	}
	if err := s.fileRepo.Create(record); err != nil {
		return nil, DatabaseError
	}

	tyreModel.Images[m.FileTypeOriginal] = record
	tyreModel.Status = m.ProcessingStatusProcessing
	if err := s.repo.Update(tyreModel); err != nil {
		return nil, fmt.Errorf("%w: %w", DatabaseError, err)
	}

	if err := s.imageProcessingService.Process(tyreModel.ID, m.FileModelTyreModel); err != nil {
		tyreModel.Status = m.ProcessingStatusFailed
		if updateErr := s.repo.Update(tyreModel); updateErr != nil {
			return nil, fmt.Errorf("processing failed: %w (also unable to mark as failed: %w)", err, updateErr)
		}
		return nil, fmt.Errorf("%w: %w", ProcessingError, err)
	}

	// Reload so the response includes the generated files
	tyreModel = s.repo.GetByID(id)

	return tyreModel, nil
}

func (s *TyreModelService) Delete(id uint) error {
	tyreModel := s.repo.GetByID(id)
	if tyreModel == nil {
		return NotFoundError
	}

	return s.repo.Delete(tyreModel)
}
